# Minimização Multidimensional


$$f(x,y)=xy$$

$$\overline{X_{n+1}} = \overline{X_n} - \nabla f_{xn}\lambda$$

In [3]:
def func1(x,y):
    return x*y

def dfx1(x,y):
    return y

def dfy1(x,y):
    return x

In [41]:
def gradient(x0,y0,xf,yf,h0,f,dfx,dfy,N=20,info=True):
    x = x0
    y = y0
    h = h0
    
    x_list = [x]
    y_list = [y]
    
    if info:
        print("\niteration 0")
        print("x:",x)
        print("y:",y)
        print("f(x,y):",f(x,y))
    
    for i in range(1,N+1):
        xn = x - dfx(x,y)*h
        yn = y - dfy(x,y)*h
        
        if (f(xn,yn)<f(x,y)):
            h *= 2
            x = xn
            y = yn
            x_list.append(x)
            y_list.append(y)
            if info:
                print("\niteration",i)
                print("x:",x)
                print("y:",y)
                print("f(x,y):",f(x,y))
        else:
            h /= 2
            
        if x>xf and y>yf:
            break
        
    return (x,y)
    

In [49]:
gradient(-5,-5,5,5,0.01,func1,dfx1,dfy1,info=True,N=10)



iteration 0
x: -5
y: -5
f(x,y): 25

iteration 1
x: -4.95
y: -4.95
f(x,y): 24.5025

iteration 2
x: -4.851
y: -4.851
f(x,y): 23.532201

iteration 3
x: -4.65696
y: -4.65696
f(x,y): 21.687276441599998

iteration 4
x: -4.2844032
y: -4.2844032
f(x,y): 18.356110780170237

iteration 5
x: -3.598898688
y: -3.598898688
f(x,y): 12.95207176648812

iteration 6
x: -2.4472511078399997
y: -2.4472511078399997
f(x,y): 5.9890379848241055

iteration 7
x: -0.8810103988223998
y: -0.8810103988223998
f(x,y): 0.776179322833204

iteration 8
x: 0.246682911670272
y: 0.246682911670272
f(x,y): 0.060852458910123225

iteration 10
x: -0.06907121526767618
y: -0.06907121526767618
f(x,y): 0.0047708327785536625


(-0.06907121526767618, -0.06907121526767618)

Como podemos ver a função tem vindo a decrescer.  
Num intervalo com $-5<x<5$ e $-5<y<5$,  
com um passo inicial de $h=0.01$ e 10 iterações:

o mínimo da função está em $(-0.06907, -0.06907)$.